test/spec/test_spec.rb

```diff
@@ -0,0 +1,67 @@
+require_relative "spec_helper"
+
+describe "A PHP application" do
+
+  it "works with the getting started guide" do
+    Hatchet::Runner.new("php-getting-started").tap do |app|
+      app.deploy do
+      # deploy works
+      end
+    end
+  end
+
+  it "checks for bad version" do
+    Hatchet::Runner.new("php-getting-started", allow_failure: true).tap do |app|
```

Comment:
Same as above with `new_app_with_stack_and_platrepo`.

Also, do not use `allow_failure` here, but instead explicitly assert a failure inside the test body, to ensure that that particular failing behavior happens, and not something completely different.

The purpose of `allow_failure` is to have a test in place that ensures future behavior, but doesn't fail for the moment until e.g. the bug in question is fixed (you'd use that for "known issues" etc).

The problem is that if the `expect()` further down does not succeed (because e.g. it accepts that version given constraint), the test would still pass.

Reply:
> do not use allow_failure here, but instead explicitly assert a failure inside the test body, to ensure that that particular failing behavior happens, and not something completely different.

This won't work if you don't pass in `allow_failure` here. By default hatchet will raise an error:

```
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> PHP app detected
remote: -----> Bootstrapping...
remote:
remote:
remote:   [UnexpectedValueException]
remote:   Could not parse version constraint 7.badversion: Invalid version string "7.badversion"
remote:
remote:
remote: config [-g|--global] [-e|--editor] [-a|--auth] [--unset] [-l|--list] [-f|--file FILE] [--absolute] [--] [<setting-key>] [<setting-value>]...
remote:
remote:
remote:  !     ERROR: An unknown internal error occurred.
remote:  !
remote:  !     Contact Heroku Support for assistance if this problem persists.
remote:  !
remote:  !     Stack trace follows for debugging purposes:
remote:  !     254 main /tmp/codon/tmp/buildpacks/d91a5b848ec0d2628b25b3bfe70a059132501cef/bin/compile
remote:
remote:  !     Push rejected, failed to compile PHP app.
remote:
remote:  !     Push failed
remote: Verifying deploy...
remote:
remote: !	Push rejected to hatchet-t-79874d6deb.
remote:
To https://git.heroku.com/hatchet-t-79874d6deb.git
 ! [remote rejected] HEAD -> main (pre-receive hook declined)
error: failed to push some refs to 'https://git.heroku.com/hatchet-t-79874d6deb.git'
Destroying "hatchet-t-79874d6deb": f2790ef2-3566-4acb-9b02-b43d3425d839, total_app_count: 274, hatchet_app_count: 20/20, finished: 0, unfinished: 20
```
 
> The problem is that if the expect() further down does not succeed (because e.g. it accepts that version given constraint), the test would still pass.

If accepts the version constraint then it would not output "Could not parse version constraint 7.badversion", no?